plugin/ipam/cni.go

```diff
@@ -35,31 +36,76 @@ func (i *Ipam) Allocate(args *skel.CmdArgs) (types.Result, error) {
 	if containerID == "" {
 		return nil, fmt.Errorf("Weave CNI Allocate: blank container name")
 	}
-	var ipnet *net.IPNet
 
-	if conf.Subnet == "" {
-		ipnet, err = i.weave.AllocateIP(containerID, false)
+	var ipconfigs []*current.IPConfig
+
+	if len(conf.IPs) > 0 {
+		// configuration includes desired IPs
+
+		var ips []net.IP
+		for _, ip := range conf.IPs {
+			ipv4 := net.ParseIP(ip).To4()
+
+			if ipv4 == nil {
+				return nil, errors.New("allocation of ipv6 addresses is not implemented")
+			}
+
+			ips = append(ips, ipv4)
+		}
+
+		for j := range ips {
+			ipnet := &net.IPNet{
+				IP:   ips[j],
+				Mask: ips[j].DefaultMask(),
+			}
+
+			err := i.weave.ClaimIP(containerID, ipnet, false)
+			if err != nil {
+				return nil, err
+			}
+
+			ipconfigs = append(ipconfigs, &current.IPConfig{
+				Version: "4",
+				Address: *ipnet,
+				Gateway: conf.Gateway,
+			})
+		}
+	} else if conf.Subnet == "" {
+		// configuration doesn't include Subnet or IPs => allocate a random IP
```

Comment:
I don't like "random" here: the allocation is quite deterministic.
How about "ask the allocator for an IP" ?

Reply:
Fixed.